llarp/util/status.hpp

```diff
@@ -0,0 +1,63 @@
+#ifndef LLARP_UTIL_STATUS_HPP
+#define LLARP_UTIL_STATUS_HPP
+#ifdef USE_ABYSS
+#include <abyss/json.hpp>
+#endif
+#include <vector>
+#include <string>
+
+namespace llarp
+{
+  namespace util
+  {
+#ifdef USE_ABYSS
+    using StatusObject_Impl = ::abyss::json::Document;
+    using Value_t           = ::abyss::json::Value;
+#else
+    struct StatusObject_Impl
+    {
+    };
+    struct Value_t
+    {
+    };
+#endif
+
+    struct StatusObject
+    {
+      StatusObject();
+      ~StatusObject();
+
+      void
+      PutInt(const char* name, uint64_t val);
+
+      void
+      PutString(const char* name, const std::string& val);
+
+      void
+      PutBool(const char* name, bool val);
+
+      void
+      PutObject(const char* name, StatusObject& obj);
+
+      void
+      PutStringArray(const char* name, std::vector< std::string >& arr);
+
+      void
+      PutObjectArray(const char* name, std::vector< StatusObject >& arr);
+
+      StatusObject_Impl Impl;
```

Comment:
DYC:

```c++
struct StatusObject
{
  using Variant = std::variant<uint64_t, std::string, bool, std::unique_ptr< StatusObject >, std::vector< std::string >, std::vector< StatusObject > >;
  using value_type = std::tuple<std::string_view, Variant>;

  void put(std::string_view name, const Variant& value);
 
  void put (const value_type& value);
  
  StatusObject();
  StatusObject(std::initializer_list<value_type> init) { /* magic */ }
};
```

then you should be able to initialize like:

```c++
StatusObject obj{{"val1", "my string"}, {"val2", false}, {"val3", {"one", "two", "three"}}};
```

Reply:
the internal implementation is optional on compile time, i.e. if we don't want to compile with jsonrpc, would this pattern allow for that?

Reply:
It should do.